js/i18n.js

```diff
@@ -1,29 +1,36 @@
-/*
- * vim: ts=4:sw=4:expandtab
- */
-;(function() {
-    'use strict';
+/* eslint-env node */
 
-    // preload.js loads this, pulling it from main.js (where it was loaded from disk)
-    var messages = window.config.localeMessages;
-    var locale = window.config.locale;
+'use strict';
 
-    window.i18n = function (message, substitutions) {
-      if (!messages[message]) {
-        return;
-      }
-      var s = messages[message].message;
-      if (substitutions instanceof Array) {
-        substitutions.forEach(function(sub) {
-          s = s.replace(/\$.+?\$/, sub);
-        });
-      } else if (substitutions) {
-        s = s.replace(/\$.+?\$/, substitutions);
-      }
-      return s;
-    };
+exports.setup = (locale, messages) => {
+  if (!locale) {
+    throw new Error('i18n: locale parameter is required');
+  }
+  if (!messages) {
+    throw new Error('i18n: messages parameter is required');
+  }
 
-    i18n.getLocale = function() {
-      return locale;
-    };
-})();
+  function getMessage(key, substitutions) {
+    const entry = messages[key];
+    if (!entry) {
+      console.error(`i18n: Attempted to get translation for nonexistent key '${key}'`);
+      return '';
+    }
+
+    const { message } = entry;
+    if (substitutions instanceof Array) {
```

Comment:
Fun fact: I tend to use `Array.isArray` having been bitten by the iframe issue in the past: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/isArray#instanceof_vs_isArray 😉